prov/efa/src/efa_cq.c

```diff
@@ -371,14 +371,212 @@ const char *efa_cq_strerror(struct fid_cq *cq_fid, int prov_errno,
 		: efa_strerror(prov_errno);
 }
 
+/**
+ * @brief Try to wait on CQ - check if ready for blocking
+ *
+ * @param[in] cq EFA completion queue
+ * @return 0 if ready to wait, -FI_EAGAIN if completions available
+ */
+int efa_cq_trywait(struct efa_cq *cq)
+{
+	struct ibv_cq *ibv_cq;
+	void *context;
+	int ret;
+
+	if (!cq->ibv_cq.channel) {
+		EFA_WARN(FI_LOG_CQ, "No ibv_comp_channel associated with CQ\n");
+		return -FI_EINVAL;
+	}
+
+	if (!ofi_cirque_isempty(cq->util_cq.cirq)) {
+		EFA_INFO(FI_LOG_CQ, "efa_cq_trywait: completions available in "
+				 "util_cq, return -FI_EAGAIN\n");
+		return -FI_EAGAIN;
+	}
+
+	ibv_cq = ibv_cq_ex_to_cq(cq->ibv_cq.ibv_cq_ex);
+	while (!ibv_get_cq_event(cq->ibv_cq.channel, &ibv_cq, &context)) {
+		ofi_atomic_inc32(&cq->nevents);
+		EFA_DBG(FI_LOG_CQ,
+				"efa_cq_trywait: drained pending event, nevents=%d\n",
+				ofi_atomic_get32(&cq->nevents));
+	}
+
+	/* Arm the completion notification for next completion */
+	ret = ibv_req_notify_cq(ibv_cq, 0);
+	if (ret) {
+		ret = -errno;
+		EFA_WARN(FI_LOG_CQ,
+				 "ibv_req_notify_cq failed with errno=%d (%s)\n", errno,
+				 fi_strerror(ret));
+		return ret;
+	}
+
+	/* Fetch any completions that we might have missed while rearming */
+	efa_cq_progress(&cq->util_cq);
+
+	return ofi_cirque_isempty(cq->util_cq.cirq) ? FI_SUCCESS : -FI_EAGAIN;
+}
+
+/**
+ * @brief Poll for completion events with timeout.
+ * 
+ * ibv_get_cq_event() waits for the next completion event in the
+ * completion event channel. Fills the arguments cq with the
+ * CQ that got the event and cq_context with the CQ's context.
+ *
+ * @param[in] cq EFA completion queue
+ * @param[in] timeout Timeout in milliseconds
+ * @return 0 on success, -FI_EAGAIN on timeout, negative error code on failure
+ */
+int efa_poll_events(struct efa_cq *cq, int timeout)
+{
+	int ret, rc;
+	void *context;
+	struct pollfd fds[2];
+	struct ibv_cq *ibv_cq;
+
+	fds[0].fd = cq->ibv_cq.channel->fd;
+	/* A signal FD for interrupting the wait operation */
+	fds[1].fd = fd_signal_get(&cq->signal);
+
+	fds[0].events = fds[1].events = POLLIN;
+
+	rc = poll(fds, 2, timeout);
+	if (rc == 0) {
+		EFA_DBG(FI_LOG_CQ, "efa_poll_events: poll() timeout after %d ms\n", timeout);
+		return -FI_EAGAIN;
+	} else if (rc < 0) {
+		ret = -errno;
+		EFA_WARN(FI_LOG_CQ,
+				 "efa_poll_events: poll() failed with errno=%d (%s)\n",
+				 errno, fi_strerror(ret));
+		return ret;
+	}
+
+	ret = 0;
+	if (fds[0].revents & POLLIN) {
+		/* Data is available for reading on completion channel */
+		ibv_cq = ibv_cq_ex_to_cq(cq->ibv_cq.ibv_cq_ex);
+		ret = ibv_get_cq_event(cq->ibv_cq.channel, &ibv_cq, &context);
+		if (ret) {
+			EFA_WARN(FI_LOG_CQ, "ibv_get_cq_event failed with ret=%d\n", ret);
+			return ret;
```

Comment:
Is there possibility of getting the fds[1] event as well on this branch?
If so is it OK we have consumed it without calling `fd_signal_reset()`?

Reply:
AFAICT the fds[1] will be reset by L471 block even if this branch happens. The only early return happens when we hit an error, which I think it's fair.